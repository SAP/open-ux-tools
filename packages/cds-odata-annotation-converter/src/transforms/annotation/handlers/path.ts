import type { Path, Identifier, Separator } from '@sap-ux/cds-annotation-parser';
import {
    PATH_TYPE,
    nodeRange,
    copyRange,
    copyPosition,
    IDENTIFIER_TYPE,
    SEPARATOR_TYPE
} from '@sap-ux/cds-annotation-parser';

import type { Element, ODataPathSeparatorDiagnostic } from '@sap-ux/odata-annotation-core-types';
import {
    createElementNode,
    createTextNode,
    Range,
    Edm,
    Diagnostic,
    DiagnosticSeverity,
    ODATA_PATH_SEPARATOR_RULE
} from '@sap-ux/odata-annotation-core-types';

import { i18n } from '../../../i18n';

import type { NodeHandler } from '../handler';
import type { VisitorState } from '../visitor-state';

import { pathLikeTypeElementName } from '../path-utils';

export const pathHandler: NodeHandler<Path> = {
    type: PATH_TYPE,
    convert(state: VisitorState, node: Path): Element | undefined {
        const elementName =
            (state.context.valueType ? pathLikeTypeElementName(state.context.valueType) : undefined) ?? Edm.Path;

        const pathText = node.value
            .split('@')
            .map((segment, i) => (i === 0 ? segment.replace(/\./g, '/') : segment))
            .join('@');

        const fragmentRanges = calculateFragmentRanges(state, node);

        state.addPath(pathText);

        const element: Element = createElementNode({
            name: elementName,
            range: nodeRange(node, true),
            contentRange: nodeRange(node, false),
            content: [createTextNode(pathText, nodeRange(node, false), fragmentRanges)]
        });
        return element;
    }
};

interface Fragment {
    image: string;
    range: Range;
}
/**
 * Fragment ranges are calculated as following.
 * before first @ all '.' chars have been replaced by '/', hence every segment considered as fragment.
 * after first @ fragments are parts separated by '/', i.e. dots are not separators of fragments in that region.
 *
 * @param state - The visitor state.
 * @param node - The path node containing segments and separators.
 * @returns An array of fragment ranges derived from the positions of segments and separators.
 */
function calculateFragmentRanges(state: VisitorState, node: Path): Range[] {
    const tokens = [...node.segments, ...node.separators].sort((a, b) => {
        const startA = a.range?.start?.character;
        const startB = b.range?.start?.character;
        if (startA && startB && startA !== startB) {
            return startA - startB;
        }
        const endA = a.range?.end?.character;
        const endB = b.range?.end?.character;
        if (endA && endB) {
            return endA - endB;
        }
        return 0;
    });

    const fragments = convertToFragments(state, node, tokens);

    return fragments.map((fragment): Range => fragment.range);
}

/**
 * Converts the segments of a Path node into Fragments, handling both data model segments and annotation references.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param node - The Path node representing the path.
 * @param tokens - The array of Identifier or Separator tokens representing the path segments.
 * @returns An array of Fragments representing the converted segments of the path.
 */
function convertToFragments(state: VisitorState, node: Path, tokens: (Identifier | Separator)[]): Fragment[] {
    // If any of the token range is not defined,
    // then we can't generate fragment ranges correctly => no ranges are returned
    const fragments: Fragment[] = [];
    let i = 0;

    // Process data model segments (until first annotation reference starting with @)
    for (; i < tokens.length; i++) {
        const token = tokens[i];
        if (token.type === IDENTIFIER_TYPE) {
            if (token.value.includes('@')) {
                // segment referencing annotation is found
                break;
            }
            if (!token.range) {
                return [];
            }

            const fragment: Fragment = {
                image: token.value,
                range: copyRange(token.range)
            };
            fragments.push(fragment);
        } else if (token.value === '/' && !token.escaped) {
            pushWrongPathSeparatorDiagnostic(state, token);
        }
    }

    const annotationFragments = processAnnotationSegments(state, node, tokens, i);
    // TODO: check if we should actually include fragments with 0 length
    // filtering fragments with no content to preserve existing behavior
    return [...fragments, ...annotationFragments].filter((fragment) => fragment.image.length);
}

/**
 * Processes the segments of an annotation path, extracting fragments and checking for missing path escaping.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param node - The Path node representing the path.
 * @param tokens - The array of Identifier or Separator tokens representing the path segments.
 * @param startIndex - The index indicating the start of the range in the tokens array.
 * @returns An array of Fragments representing the processed segments of the annotation path.
 */
function processAnnotationSegments(
    state: VisitorState,
    node: Path,
    tokens: (Identifier | Separator)[],
    startIndex: number
): Fragment[] {
    const fragments: Fragment[] = [];
    let requiresEscaping = false;
    let i = startIndex;
    while (i < tokens.length) {
        let token = tokens[i];

        if (!token.range) {
            return [];
        }

        const fragment: Fragment = {
            image: '',
            range: Range.create(copyPosition(token.range.start), copyPosition(token.range.start))
        };

        while (token?.type === IDENTIFIER_TYPE || token?.value === '.') {
            fragment.image += token.value;

            if (fragment.range && token.range) {
                fragment.range.end = copyPosition(token.range.end);
            }

            i++;
            token = tokens[i];

            if (token?.type === SEPARATOR_TYPE && token.value === '/' && !token.escaped) {
                requiresEscaping = true;
            }
        }

        fragments.push(fragment);
        i++;
    }

    if (requiresEscaping) {
        pushMissingPathEscapingDiagnostic(state, node, tokens, startIndex);
    }

    return fragments;
}

/**
 * Pushes a diagnostic for missing path escaping brackets in the specified range of tokens.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param node - The Path node representing the path.
 * @param tokens - The array of Identifier or Separator tokens representing the path segments.
 * @param startIndex - The index indicating the start of the range in the tokens array.
 */
function pushMissingPathEscapingDiagnostic(
    state: VisitorState,
    node: Path,
    tokens: (Identifier | Separator)[],
    startIndex: number
): void {
    const start = tokens[startIndex];
    const end = tokens.slice(-1)[0];
    if (!start?.range || !end?.range) {
        return;
    }

    const pathSeparatorDiagnostic: ODataPathSeparatorDiagnostic = {
        rule: ODATA_PATH_SEPARATOR_RULE,
        message: i18n.t('Path_escaping_brackets_should_be_used'),
        range: Range.create(copyPosition(start.range.start), copyPosition(end.range.end)),
        severity: DiagnosticSeverity.Warning
    };

    const valueToBeEscaped = '@' + node.value.split('@').slice(1).join('@');

    const prefix = start.type === IDENTIFIER_TYPE && start.quoted ? '' : '![';
    const suffix = ']';
    // If there are existing escape sequences, we need to remove them
    const body = valueToBeEscaped.replace(/[![\]]/g, '');

    pathSeparatorDiagnostic.data = {
        value: valueToBeEscaped,
        proposedValue: prefix + body + suffix
    };

    state.addDiagnostic(pathSeparatorDiagnostic);
}

/**
 * Pushes a diagnostic for a wrong path separator in the given token range.
 *
 * @param state - The VisitorState object managing the state during the visit.
 * @param token - The Separator token representing the path separator.
 */
function pushWrongPathSeparatorDiagnostic(state: VisitorState, token: Separator): void {
    if (!token.range) {
        return;
    }
    // code action to correct wrong separator
    const message = i18n.t('Wrong_path_separator_0_Did_you_mean_1', {
        currentValue: '/',
        proposedValue: '.',
        interpolation: { escapeValue: false }
    });

    const diagnostic = Diagnostic.create(token.range, message, DiagnosticSeverity.Warning);
    diagnostic.data = {
        caseCheck: {
            value: '/',
            proposedValue: '.',
            lookupPath: [],
            isNamespacedValue: false
        }
    };
    state.addDiagnostic(diagnostic);
}
