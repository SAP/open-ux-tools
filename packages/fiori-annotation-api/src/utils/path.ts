import { realpathSync } from 'fs';
import { fileURLToPath } from 'url';

const driveLetter = process.platform === 'win32' ? realpathSync.native('\\')[0] : '';

/**
 *  Converts URI to path.
 *
 * @param uri - URI.
 * @returns File path.
 */
export function pathFromUri(uri: string): string {
    const parsedUri = fileURLToPath(uri);

    // for windows, some NodeJS methods will output uppercase drive letters, some in lowercase
    if (process.platform === 'win32') {
        return toggleCase(parsedUri.charAt(0)) + parsedUri.slice(1);
    }

    return parsedUri;
}

/**
 * Changes the drive letter to the same as `realpathSync`.
 *
 * @param character Drive letter character
 * @returns Normalized drive letter character
 */
function toggleCase(character: string): string {
    return driveLetter === driveLetter.toUpperCase() ? character.toUpperCase() : character.toLowerCase();
}
