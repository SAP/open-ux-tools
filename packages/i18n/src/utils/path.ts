import { stat } from 'fs';
import type { CdsEnvironment } from '../types';
import { getI18nConfiguration } from './config';

/**
 * Get json path.
 *
 * @param path file path
 * @returns .json file path
 */
export function jsonPath(path: string): string {
    return `${path}.json`;
}
/**
 * Get properties path.
 *
 * @param path file path
 * @param env cds environment
 * @returns .properties file path
 */
export function capPropertiesPath(path: string, env: CdsEnvironment): string {
    const { fallbackLanguage } = getI18nConfiguration(env);
    const languageSuffix = fallbackLanguage === '' ? '' : `_${fallbackLanguage}`;
    return `${path}${languageSuffix}.properties`;
}

/**
 * Get csv path.
 *
 * @param path file path
 * @returns .csv file path
 */
export function csvPath(path: string): string {
    return `${path}.csv`;
}

/**
 * Check if a folder of a file exists.
 *
 * @param path an absolute path to a folder or a file
 * @returns boolean
 */
export function doesExist(path: string): Promise<boolean> {
    return new Promise((resolve) => {
        stat(path, (err) => {
            if (err) {
                resolve(false);
            } else {
                resolve(true);
            }
        });
    });
}
