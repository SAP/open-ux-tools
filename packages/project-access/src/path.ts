import { realpathSync } from 'fs';
import { fileURLToPath } from 'url';

const driveLetter = process.platform === 'win32' ? realpathSync.native('\\')[0] : '';

/**
 *  Normalize URI to path.
 *
 * @param pathOrUri - URI or string.
 * @param parse - URI or string.
 * @returns File path.
 */
export function normalizePath(pathOrUri: string, parse = true): string {
    const parsedPath = parse ? fileURLToPath(pathOrUri) : pathOrUri;

    // for windows, some NodeJS methods will output uppercase drive letters, some in lowercase
    if (process.platform === 'win32') {
        return toggleCase(parsedPath.charAt(0)) + parsedPath.slice(1);
    }

    return parsedPath;
}

/**
 * Changes the drive letter to the same as `realpathSync`.
 *
 * @param character Drive letter character
 * @returns Normalized drive letter character
 */
function toggleCase(character: string): string {
    return driveLetter === driveLetter.toUpperCase() ? character.toUpperCase() : character.toLowerCase();
}
