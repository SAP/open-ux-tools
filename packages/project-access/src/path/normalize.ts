import { realpathSync } from 'node:fs';

let toUpperCase: boolean;

/**
 *  Normalize path to canonical form.
 *
 * @param path - URI or string.
 * @returns File path.
 */
export function normalizePath(path: string): string {
    // for windows, some NodeJS methods will output uppercase drive letters, some in lowercase
    if (process.platform === 'win32') {
        if (toUpperCase === undefined) {
            const driveLetter = realpathSync.native('\\')[0];
            toUpperCase = driveLetter === driveLetter.toUpperCase();
        }
        const correctedDriveLetter = toUpperCase ? path.charAt(0).toUpperCase() : path.charAt(0).toLowerCase();
        return correctedDriveLetter + path.slice(1);
    }

    return path;
}
