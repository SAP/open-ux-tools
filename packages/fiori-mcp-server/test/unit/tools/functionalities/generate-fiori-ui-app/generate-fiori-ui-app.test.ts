import { join } from 'path';
const mockFindInstalledPackages = jest.fn().mockResolvedValue([
    {
        path: 'node_modules/@sap/generator-fiori',
        /** Path to the package.json */
        packageJsonPath: 'node_modules/@sap/generator-fiori/package.json',
        /** The parsed package info */
        packageInfo: {
            name: '@sap/generator-fiori',
            version: '1.18.5'
        }
    }
]);
jest.mock('@sap-ux/nodejs-utils', () => ({
    findInstalledPackages: mockFindInstalledPackages
}));

import {
    GENERATE_FIORI_UI_APP,
    generateFioriUIAppHandlers
} from '../../../../../src/tools/functionalities/generate-fiori-ui-app';
import { existsSync } from 'fs';

// Mock child_process.exec
const mockExec = jest.fn();
const testOutputDir = join(__dirname, '../../../../test-output/');

jest.mock('child_process', () => ({
    exec: (...args: any) => mockExec(...args)
}));

describe('getFunctionalityDetails', () => {
    test('getFunctionalityDetails', async () => {
        const details = await generateFioriUIAppHandlers.getFunctionalityDetails({
            appPath: join(testOutputDir, 'app1'),
            functionalityId: GENERATE_FIORI_UI_APP.id
        });
        expect(details).toEqual(GENERATE_FIORI_UI_APP);
    });
});
const paramTest = {
    projectPath: join(testOutputDir, 'app1'),
    appGenConfig: {
        version: '1.0.0',
        floorplan: 'FE_LROP',
        project: {
            name: 'app1',
            targetFolder: join(testOutputDir, 'app1'),
            namespace: 'zzz',
            title: 'App 1',
            description: 'Description for App 1',
            ui5Theme: 'sap_belize',
            ui5Version: '1.100.0',
            localUI5Version: '1.100.0',
            sapux: true,
            skipAnnotations: false,
            enableCodeAssist: true,
            enableEslint: true,
            enableTypeScript: true
        },
        service: {
            capService: {
                projectPath: 'zzzapp1',
                serviceName: 'app1',
                serviceCdsPath: 'srv/cat-service.cds',
                capType: 'Node.js' // optional
            },
            servicePath: 'app1'
        },
        entityConfig: {
            mainEntity: {
                entityName: 'Travel'
            },
            generateFormAnnotations: true,
            generateLROPAnnotations: true
        },
        telemetryData: {
            generationSourceName: 'test',
            generationSourceVersion: '1.0.0'
        }
    }
};

describe('executeFunctionality', () => {
    test('executeFunctionality - success', async () => {
        mockExec.mockImplementation((cmd, opts, callback) => {
            callback(null, 'mock stdout', 'mock stderr');
        });
        const result = await generateFioriUIAppHandlers.executeFunctionality({
            appPath: join(testOutputDir, 'app1'),
            functionalityId: GENERATE_FIORI_UI_APP.id,
            parameters: paramTest
        });
        expect(result).toEqual(
            expect.objectContaining({
                appPath: join(testOutputDir, 'app1/app/app1'),
                changes: [],
                functionalityId: 'generate-fiori-ui-app',
                message: `Generation completed successfully: ${join(testOutputDir, 'app1/app/app1')}`,
                parameters: {
                    appGenConfig: {
                        version: '1.0.0',
                        floorplan: 'FE_LROP',
                        project: {
                            description: 'Description for App 1',
                            enableCodeAssist: true,
                            enableEslint: true,
                            enableTypeScript: true,
                            localUI5Version: '1.100.0',
                            name: 'app1',
                            'namespace': 'zzz',
                            'sapux': true,
                            'skipAnnotations': false,
                            targetFolder: join(testOutputDir, 'app1'),
                            'title': 'App 1',
                            'ui5Theme': 'sap_belize',
                            'ui5Version': '1.100.0'
                        },
                        service: {
                            servicePath: 'app1',
                            capService: {
                                serviceName: 'app1',
                                'capType': 'Node.js',
                                'projectPath': 'zzzapp1',
                                'serviceCdsPath': 'srv/cat-service.cds'
                            }
                        },
                        entityConfig: {
                            mainEntity: {
                                entityName: 'Travel'
                            },
                            generateFormAnnotations: true,
                            generateLROPAnnotations: true
                        },
                        telemetryData: {
                            generationSourceName: 'test',
                            generationSourceVersion: '1.0.0'
                        }
                    },
                    projectPath: join(testOutputDir, 'app1')
                },
                status: 'Success'
            })
        );
        expect(existsSync(join(testOutputDir, 'app1', 'default-generator-config.json'))).toBeFalsy();
    });

    test('executeFunctionality - unsuccess', async () => {
        mockExec.mockImplementation((cmd, opts, callback) => {
            throw new Error('Dummy');
        });
        const result = await generateFioriUIAppHandlers.executeFunctionality({
            appPath: join(testOutputDir, 'app1'),
            functionalityId: GENERATE_FIORI_UI_APP.id,
            parameters: paramTest
        });
        expect(result).toEqual(
            expect.objectContaining({
                appPath: join(testOutputDir, 'app1/app/app1'),
                changes: [],
                functionalityId: 'generate-fiori-ui-app',
                message: `Error generating application: Dummy`,
                parameters: paramTest,
                status: 'Error'
            })
        );
        expect(existsSync(join(testOutputDir, 'app1', 'default-generator-config.json'))).toBeFalsy();
    });

    test('executeFunctionality - empty parameters', async () => {
        mockExec.mockImplementation((cmd, opts, callback) => {
            throw new Error('Dummy');
        });
        await expect(
            generateFioriUIAppHandlers.executeFunctionality({
                appPath: '',
                functionalityId: GENERATE_FIORI_UI_APP.id,
                parameters: {}
            })
        ).rejects.toThrowErrorMatchingInlineSnapshot(`
            "Missing required fields in generatorConfig. [
                {
                    \\"expected\\": \\"object\\",
                    \\"code\\": \\"invalid_type\\",
                    \\"path\\": [
                        \\"appGenConfig\\"
                    ],
                    \\"message\\": \\"Invalid input: expected object, received undefined\\"
                }
            ]"
        `);
    });

    test('executeFunctionality - wrong appGenConfig', async () => {
        mockExec.mockImplementation((cmd, opts, callback) => {
            throw new Error('Dummy');
        });
        await expect(
            generateFioriUIAppHandlers.executeFunctionality({
                appPath: 'app1',
                functionalityId: GENERATE_FIORI_UI_APP.id,
                parameters: {
                    projectPath: 'app1',
                    appGenConfig: 'dummy'
                }
            })
        ).rejects.toThrowErrorMatchingInlineSnapshot(`
            "Missing required fields in generatorConfig. [
                {
                    \\"expected\\": \\"object\\",
                    \\"code\\": \\"invalid_type\\",
                    \\"path\\": [
                        \\"appGenConfig\\"
                    ],
                    \\"message\\": \\"Invalid input: expected object, received string\\"
                }
            ]"
        `);
    });
});
