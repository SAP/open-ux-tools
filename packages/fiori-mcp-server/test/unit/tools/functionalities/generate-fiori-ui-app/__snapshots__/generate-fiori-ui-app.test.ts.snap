// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getFunctionalityDetails getFunctionalityDetails 1`] = `
Object {
  "description": "Creates (generates) a new SAP Fiori UI application within an existing CAP project.
                To populate parameters, you **MUST** use the ***CDS MCP*** to search the model for service definitions, entities, associations, and UI annotations.
                As a fallback, only if no such tool is available, you should manually read and parse all .cds files in the projectPath to extract this information.
                The data obtained from either method must then be formatted into a JSON object and passed as the parameters.
                The configuration **MUST** be a valid JSON object corresponding to the inputSchema of the tool.
                The configuration **MUST** be based on the project files in the projectPath.",
  "functionalityId": "generate-fiori-ui-app",
  "name": "Generate SAP Fiori UI Application",
  "parameters": Object {
    "additionalProperties": false,
    "description": "The configuration that will be used for the Application UI generation.
            The configuration **MUST** be a valid JSON object corresponding to the inputSchema of the functionality.",
    "properties": Object {
      "entityConfig": Object {
        "additionalProperties": false,
        "properties": Object {
          "generateFormAnnotations": Object {
            "type": "boolean",
          },
          "generateLROPAnnotations": Object {
            "type": "boolean",
          },
          "mainEntity": Object {
            "additionalProperties": false,
            "properties": Object {
              "entityName": Object {
                "description": "The name of the main entity. EntitySet Name attribute in OData Metadata",
                "examples": Array [
                  "'SalesOrder'",
                  "'PurchaseOrderHeader'",
                  "'MyEntity'",
                ],
                "type": "string",
              },
            },
            "required": Array [
              "entityName",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "mainEntity",
          "generateFormAnnotations",
          "generateLROPAnnotations",
        ],
        "type": "object",
      },
      "floorplan": Object {
        "description": "SAP Fiori Elements floor plan type.",
        "enum": Array [
          "FE_FPM",
          "FE_LROP",
          "FE_OVP",
          "FE_ALP",
          "FE_FEOP",
          "FE_WORKLIST",
          "FF_SIMPLE",
        ],
        "type": "string",
      },
      "project": Object {
        "additionalProperties": false,
        "properties": Object {
          "description": Object {
            "type": "string",
          },
          "name": Object {
            "pattern": "^[a-z0-9-]+$",
            "type": "string",
          },
          "sapux": Object {
            "default": true,
            "type": "boolean",
          },
          "targetFolder": Object {
            "description": "Absolute path to the project folder (projectPath).",
            "type": "string",
          },
          "title": Object {
            "type": "string",
          },
          "ui5Version": Object {
            "default": "1.136.7",
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "description",
          "targetFolder",
          "ui5Version",
          "sapux",
        ],
        "type": "object",
      },
      "projectType": Object {
        "description": "SAP Fiori Elements project type. Corresponds to the SAP Fiori Elements floor plan.",
        "enum": Array [
          "LIST_REPORT_OBJECT_PAGE",
          "FORM_ENTRY_OBJECT_PAGE",
          "FLEXIBLE_PROGRAMMING_MODEL",
        ],
        "type": "string",
      },
      "service": Object {
        "additionalProperties": false,
        "properties": Object {
          "capService": Object {
            "additionalProperties": false,
            "properties": Object {
              "capType": Object {
                "enum": Array [
                  "Node.js",
                  "Java",
                ],
                "type": "string",
              },
              "projectPath": Object {
                "type": "string",
              },
              "serviceCdsPath": Object {
                "description": "The path to the service cds file",
                "examples": Array [
                  "srv/service.cds",
                  "srv/my-service.cds",
                  "path/to/srv/service.cds",
                  "path/to/srv/my-service.cds",
                ],
                "type": "string",
              },
              "serviceName": Object {
                "type": "string",
              },
            },
            "required": Array [
              "projectPath",
              "serviceName",
              "serviceCdsPath",
            ],
            "type": "object",
          },
        },
        "required": Array [
          "capService",
        ],
        "type": "object",
      },
      "telemetryData": Object {
        "additionalProperties": false,
        "properties": Object {
          "generationSourceName": Object {
            "default": "@sap-ux/fiori-mcp-server",
            "type": "string",
          },
          "generationSourceVersion": Object {
            "default": "0.2.4",
            "type": "string",
          },
        },
        "required": Array [
          "generationSourceName",
          "generationSourceVersion",
        ],
        "type": "object",
      },
      "version": Object {
        "default": "0.2",
        "description": "Config schema version.",
        "type": "string",
      },
    },
    "required": Array [
      "entityConfig",
      "floorplan",
      "project",
      "projectType",
      "service",
      "telemetryData",
      "version",
    ],
    "type": "object",
  },
}
`;
